kdc/httpkadmind.c

```diff
@@ -1065,14 +1166,15 @@ get_keys1(kadmin_request_desc r, const char *pname)
          * We're writing, but maybe we can't (not a primary) or have to
          * reconnect to a primary kadmind.
          */
-        if (ret == 0 && primary_server && !primary_admin_server) {
+        if (ret == 0 && primary_server_URI && !writable_kadmin_server) {
             /* Local DB, not a primary -> redirect */
             ret = KADM5_READ_ONLY;
-        } else if (ret == 0 && primary_admin_server) {
+        } else if (ret == 0 && writable_kadmin_server) {
             /* Connected to replica -> reconnect to primary */
             kadm5_destroy(r->kadm_handle);
-            ret = get_kadm_handle(r->context, &r->kadm_handle, 1);
-        } else if (ret == 0 && local_read_only) {
+            ret = get_kadm_handle(r->context, r->realm, 1 /* want_write */,
+                                  &r->kadm_handle);
+        } else if (ret == 0 && local_hdb_read_only) {
             ret = KADM5_READ_ONLY;
         }
         princ.attributes |= KRB5_KDB_MATERIALIZE;
```

Comment:
Let's ifdef out the materialization code for now -- we need to handle the conversion of TLs to TL_EXTENSION.